File: ios/GradeThread/Accessibility/A11yAnnounce.swift
import SwiftUI
import UIKit

/// US-701 — a tiny reusable wrapper so async outcomes (grade complete, eBay
/// sync done, Scout scan finished, form saved) announce themselves to VoiceOver
/// instead of silently swapping content (WCAG 4.1.3 Status Messages). Prior to
/// this only PhotoIntakeView announced anything; every other long-running
/// result was inaudible.
enum A11yAnnounce {

    /// Posts a VoiceOver announcement. No-op when VoiceOver isn't running, so
    /// it's safe to call unconditionally from result handlers.
    @MainActor
    static func announce(_ message: String) {
        guard UIAccessibility.isVoiceOverRunning, !message.isEmpty else { return }
        // Defer a beat so the announcement lands after SwiftUI commits the
        // content change, otherwise VoiceOver can drop it.
        DispatchQueue.main.asyncAfter(deadline: .now() + 0.1) {
            AccessibilityNotification.Announcement(message).post()
        }
    }

    /// Signals that the screen's contents changed substantially, moving
    /// VoiceOver focus to the start of the new content (or to `focusing` text).
    @MainActor
    static func screenChanged(focusing message: String? = nil) {
        guard UIAccessibility.isVoiceOverRunning else { return }
        DispatchQueue.main.asyncAfter(deadline: .now() + 0.1) {
            UIAccessibility.post(notification: .screenChanged, argument: message)
        }
    }
}

File: android/app/src/main/java/com/gradethread/app/ui/a11y/A11yAnnounce.kt
package com.gradethread.app.ui.a11y

import android.content.Context
import android.view.View
import android.view.accessibility.AccessibilityEvent
import android.view.accessibility.AccessibilityManager
import androidx.compose.runtime.Composable
import androidx.compose.runtime.remember
import androidx.compose.ui.platform.LocalView

/**
 * US-1304: announcements for async outcomes (iOS A11yAnnounce, US-701 —
 * WCAG 4.1.3 Status Messages). Grade complete, sync done, form saved: results
 * that swap content silently are inaudible to TalkBack users. Call
 * [A11yAnnouncer.announce] from result handlers; it's a no-op when no
 * accessibility service is running, so call it unconditionally.
 *
 * For text that lives IN the composition (inline status lines), prefer
 * `Modifier.semantics { liveRegion = LiveRegionMode.Polite }` — this
 * imperative path is for transient outcomes with no persistent node.
 */
class A11yAnnouncer(private val view: View) {

    private val manager: AccessibilityManager? =
        view.context.getSystemService(Context.ACCESSIBILITY_SERVICE) as? AccessibilityManager

    private val enabled: Boolean get() = manager?.isEnabled == true

    /**
     * Speak [message] through the active accessibility service. Deferred a
     * beat so it lands after Compose commits the content change (the same
     * drop-avoidance delay the iOS helper uses).
     */
    fun announce(message: String) {
        if (!enabled || message.isEmpty()) return
        view.postDelayed({
            @Suppress("DEPRECATION") // the View API remains the reliable
            // route for transient, node-less announcements; the deprecation
            // steers persistent text toward liveRegion semantics (which the
            // in-composition surfaces here already use).
            view.announceForAccessibility(message)
        }, ANNOUNCE_DELAY_MS)
    }

    /**
     * Signal a substantial content change so focus moves to the new content
     * (iOS `screenChanged`).
     */
    fun screenChanged() {
        if (!enabled) return
        view.postDelayed({
            view.sendAccessibilityEvent(AccessibilityEvent.TYPE_WINDOW_STATE_CHANGED)
        }, ANNOUNCE_DELAY_MS)
    }

    private companion object {
        const val ANNOUNCE_DELAY_MS = 100L
    }
}

/** The composable entry point: `val a11y = rememberA11yAnnouncer()`. */
@Composable
fun rememberA11yAnnouncer(): A11yAnnouncer {
    val view = LocalView.current
    return remember(view) { A11yAnnouncer(view) }
}
